core/inc/knoting/input_manager.h

```diff
@@ -0,0 +1,246 @@
+    Home = 268,
+    End = 269,
+    CapsLock = 280,
+    ScrollLock = 281,
+    NumLock = 282,
+    PrintScreen = 283,
+    Pause = 284,
+    F1 = 290,
+    F2 = 291,
+    F3 = 292,
+    F4 = 293,
+    F5 = 294,
+    F6 = 295,
+    F7 = 296,
+    F8 = 297,
+    F9 = 298,
+    F10 = 299,
+    F11 = 300,
+    F12 = 301,
+    F13 = 302,
+    F14 = 303,
+    F15 = 304,
+    F16 = 305,
+    F17 = 306,
+    F18 = 307,
+    F19 = 308,
+    F20 = 309,
+    F21 = 310,
+    F22 = 311,
+    F23 = 312,
+    F24 = 313,
+    F25 = 314,
+    KP0 = 320,
+    KP1 = 321,
+    KP2 = 322,
+    KP3 = 323,
+    KP4 = 324,
+    KP5 = 325,
+    KP6 = 326,
+    KP7 = 327,
+    KP8 = 328,
+    KP9 = 329,
+    KPDecimal = 330,
+    KPDivide = 331,
+    KPMultiply = 332,
+    KPSubtract = 333,
+    KPAdd = 334,
+    KPEnter = 335,
+    KPEqual = 336,
+    LeftShift = 340,
+    LeftControl = 341,
+    LeftAlt = 342,
+    LeftSuper = 343,
+    RightShift = 344,
+    RightControl = 345,
+    RightAlt = 346,
+    RightSuper = 347,
+    Menu = 348,
+    Last
+};
+
+class input_manager {
+   public:
+    /// KeyBoard Methods
+    bool m_keyPressed(KeyCode key);
+    bool m_keyHeldDown(KeyCode key);
+    bool m_keyOnTrigger(KeyCode key);
+    bool m_keyOnRelease(KeyCode key);
+
+    /// Pad Methods
+    bool m_padButtonPressed(PadCode pad, PadButtonCode button);
+    bool m_padButtonHeldDown(PadCode pad, PadButtonCode button);
+    bool m_padButtonTriggered(PadCode pad, PadButtonCode button);
+    bool m_padButtonReleased(PadCode pad, PadButtonCode button);
+    float m_getPadAxis(PadCode pad, JoyStickCode joy_stick);
+    bool m_padPresent(PadCode pad);
+
+    /// Mouse Methods
+    bool m_mouseButtonPressed(MouseButtonCode button);
+    bool m_mouseButtonHeldDown(MouseButtonCode button);
+    bool m_mouseButtonTriggered(MouseButtonCode button);
+    bool m_mouseButtonReleased(MouseButtonCode button);
+    bool m_mouseDoubleClicked(MouseButtonCode button);
+    void SetDoubleClickLimit(float ms);
+    void m_setMouseSensitivity(float sensitivity);
+    bool m_wheelMoved();
+    vec2 m_getRelativePosition();
+    vec2 m_getAbsolutePosition();
+    vec2 m_getWheelMovement();
+
+   protected:
+    friend Window;
+
+    input_manager(Window& owner);
+    ~input_manager();
+
+    void m_updatePads(GLFWwindow* glfwWindow);
+    void m_keyEvent(int key, bool pressed);
+    void m_mouseButtonEvent(int button, bool pressed);
+    void m_mouseEvent(double x, double y);
+    void m_updateHolds();
+    void m_updateRelativePositions();
+    void m_scrollEvent(vec2 offset);
+
+    bool m_keyBinding[(int)KeyCode::Last];
+    bool m_mouseBindings[(int)MouseButtonCode::Last];
+    bool m_padBindings[(int)PadCode::Last][(int)PadButtonCode::Last];
+    bool m_holdKeyBindings[(int)KeyCode::Last];
+    bool m_holdMouseBindings[(int)MouseButtonCode::Last];
+    bool m_holdPadBindings[(int)PadCode::Last][(int)PadButtonCode::Last];
+    vec2 m_lastScroll;
+    vec<2, double> m_mousePosition;
```

Comment:
I was the one who told you for this to be `vec<2, double>` but I think it can be `vec2d`